src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -643,19 +642,24 @@ private enum ReceiverCaptureMode
             ImmutableArray<int> argsToParamsOpt,
             ImmutableArray<RefKind> argumentRefKindsOpt,
             ArrayBuilder<BoundExpression>? storesOpt,
-            ref ArrayBuilder<LocalSymbol>? tempsOpt,
+            [NotNullIfNotNull(nameof(tempsOpt))] ref ArrayBuilder<LocalSymbol>? tempsOpt,
             BoundExpression? firstRewrittenArgument = null)
         {
             Debug.Assert(argumentRefKindsOpt.IsDefault || argumentRefKindsOpt.Length == arguments.Length);
             var requiresInstanceReceiver = methodOrIndexer.RequiresInstanceReceiver() && methodOrIndexer is not MethodSymbol { MethodKind: MethodKind.Constructor } and not FunctionPointerMethodSymbol;
             Debug.Assert(!requiresInstanceReceiver || rewrittenReceiver != null || _inExpressionLambda);
             Debug.Assert(captureReceiverMode == ReceiverCaptureMode.Default || (requiresInstanceReceiver && rewrittenReceiver != null && storesOpt is object));
 
+            var adjustedReceiver = AdjustReceiverForExtensionsIfNeeded(rewrittenReceiver, methodOrIndexer, storesOpt, ref tempsOpt);
+            bool wasReceiverAdjusted = !object.ReferenceEquals(adjustedReceiver, rewrittenReceiver);
+            rewrittenReceiver = adjustedReceiver;
+
             BoundLocal? receiverTemp = null;
             BoundAssignmentOperator? assignmentToTemp = null;
 
             if (captureReceiverMode != ReceiverCaptureMode.Default ||
-                (requiresInstanceReceiver && arguments.Any(a => usesReceiver(a))))
+                (requiresInstanceReceiver && arguments.Any(a => usesReceiver(a))) ||
+                wasReceiverAdjusted)
```

Comment:
>wasReceiverAdjusted [](http://example.com/codeflow?start=16&length=19)

This looks suspicious. Why is this necessary?

Reply:
The two "Transform" callers do a `CanValueChangeBetweenReads` check and pass `ReceiveCaptureMode.Default` when the value cannot change (or other capture modes when the value could change). For example, using a constant or `this` in a struct as the receiver.
It seems unnecessary to evaluate `Unsafe.As` twice in such cases, so we detect it and capture the temp anyways.
The extra temp did not appear to have any negative effect in the normal case (no compound assignment and no interpolation receiver capture), but I debated conditioning this only for those two "Transform" scnearios, by either adding a new capture mode and/or moving the `CanValueChangeBetweenReads` check here.

Reply:
Also consider the following. And `Unsafe.As` is an intrinsic and its calls will be completely removed from the code by JIT, whereas the unnecessary (from semantics point of view) locals will likely stay. Therefore, I am not convinced that we need to worry about those extra calls.